esp/bindings/SOAP/xpp/fxpp/FragmentedXmlPullParser.h

```diff
@@ -0,0 +1,2145 @@
+        const char* path = nullptr;
+        const char* match = nullptr;
+
+        if (m_resourceAttribute.isEmpty() && m_pathAttribute.isEmpty())
+            throw XmlPullParserException("invalid detector configuration - no primary key attribute");
+        if (!m_resourceAttribute.isEmpty())
+            resource = stag.getValue(m_resourceAttribute);
+        if (!m_pathAttribute.isEmpty())
+            path = stag.getValue(m_pathAttribute);
+        if (resource && path)
+            throw XmlPullParserException("invalid markup - external fragment may be resource or path, but not both");
+        if (!resource && !path)
+            return nullptr;
+        if (!m_matchAttribute.isEmpty())
+            match = stag.getValue(m_matchAttribute);
+        if (resource)
+            return new Detection(CFragmentedXmlAssistant::Resource, resource, match);
+        return new Detection(CFragmentedXmlAssistant::Path, path, match);
+    }
+
+protected:
+    StringBuffer m_resourceAttribute;
+    StringBuffer m_pathAttribute;
+    StringBuffer m_matchAttribute;
+
+public:
+    CExternalFragmentDetector()
+    {
+        setResourceAttribute("resource");
+        setPathAttribute("path");
+        setMatchAttribute("match");
+    }
+
+    void setResourceAttribute(const char* value)
+    {
+        m_resourceAttribute.set(value);
+    }
+
+    void setPathAttribute(const char* value)
+    {
+        m_pathAttribute.set(value);
+    }
+
+    void setMatchAttribute(const char* value)
+    {
+        m_matchAttribute.set(value);
+    }
+};
+
+/**
+ * Reusable base class for `IXmlFragment` implementations that copies all fragment data passed to
+ * it, unlike CSimpleXmlFragment which which assumes the data buffers passed to it will be valid
+ * at least as long as it is.
+ */
+class CBaseXmlFragment : public CInterface
+{
+public:
+    const char*      queryUId() const { return m_uid; }
+    const char*      queryContent() const { return m_content; }
+    int              queryContentLength() const { return int(m_content.length()); }
+    XmlFragmentRules queryRules() const { return m_rules; }
+
+protected:
+    StringAttr       m_uid;
+    StringBuffer     m_content;
+    XmlFragmentRules m_rules;
+
+protected:
+    CBaseXmlFragment(const char* uid, const char* content, int length, XmlFragmentRules rules)
+        : m_uid(uid)
+        , m_content(length, content)
+        , m_rules(rules)
+    {
+    }
+
+    CBaseXmlFragment()
+        : m_rules(0)
+    {
+    }
+};
+
+/**
+ * `IXmlFragment` implementation exposing `CbaseXmlFragment` functionality.
+ */
+class CSafeXmlFragment : implements IXmlFragment, public CBaseXmlFragment
+{
+public:
+    IMPLEMENT_IINTERFACE;
+    IMPLEMENT_IXMLFRAGMENT_WITH(CBaseXmlFragment);
+
+    CSafeXmlFragment(const char* uid, const char* content, int length, XmlFragmentRules rules)
+        : CBaseXmlFragment(uid, content, length, rules)
+    {
+    }
+};
+
+class CBaseXmlFragmentGroup : public CBaseXmlFragment
```

Comment:
I would hope that once the classes are in the cpp and only the external interfaces are exposed that the concept of FragmentGroup would be internal to the cpp file and not exposed outside the black box?